src/stdx/zipfian.zig

```diff
@@ -0,0 +1,490 @@
+//!
+//! In practice these probabilities often need to be spread across e.g. a
+//! table's keyspace, which involves some kind of mapping step from index to index.
+//! Because that mapping is non-trivial to optimize, it is also provided here.
+//!
+//! The algorithm here is based on
+//! "Quickly Generating Billion-Record Synthetic Databases", Jim Gray et al, SIGMOD 1994.
+//! Per the paper it is adapted from Knuth vol 3.
+//! This is also the algorithm used by YCSB's ZipfianGenerator.java.
+//! Note that the code listing in the paper contains obvious errors,
+//! corrected here and in YCSB.
+//!
+//! There are two generators here,
+//! both of which generate random keys from 0 to a specified maximum.
+//! In the basic `ZipfianGenerator`, key 0 has the highest probability,
+//! 1 the next highest, etc.
+//! The `ShuffledZipfian` generator instead spreads the distribution out
+//! across the key space as if it were a shuffled deck.
+//!
+//! Both generators allow for the key space to grow (but not shrink),
+//! dynamically recomputing the distribution. When the `ShuffledZipfian`
+//! generator grows it acts as if each new key was inserted into the shuffled
+//! deck randomly, preserving the relative probability of existing keys.
+//!
+//! The non-shuffled generator should pass a 2-sample Kolmogorov–Smirnov test;
+//! the shuffled generator does not because the tail distribution is fudged.
+
+const std = @import("std");
+const assert = @import("std").debug.assert;
+const Random = @import("std").Random;
+const math = @import("std").math;
+const BoundedArray = @import("./bounded_array.zig").BoundedArray;
+
+/// The default "skew" of the distribution.
+const theta_default = 0.99; // per YCSB
+
+/// Generates Zipfian-distributed numbers from 0 to a specified maximum.
+///
+/// The internal variables here are the same is in the paper;
+/// the external intended to be more understandable to the user.
+pub const ZipfianGenerator = struct {
+    const Self = @This();
+
+    theta: f64,
+
+    /// The number of items in the set.
+    n: u64,
+    /// The Riemann zeta function calculated up to `n`,
+    /// aka the "generalized harmonic number" of order `theta` for `n`.
+    /// This is a pre-calculated factor in the probability of any particular item
+    /// being selected.
+    /// It is expensive to calculate for large but useful values of `n`,
+    /// but can be calculated incrementally as `n` grows.
+    zetan: f64,
+
+    /// Create a generator from `[0, items)` with `theta` equal to 0.99.
+    pub fn init(items: u64) ZipfianGenerator {
+        return ZipfianGenerator.init_theta(items, theta_default);
+    }
+
+    /// Create a generator from `[0, items)` with given `theta`.
+    ///
+    /// `theta` is the "skew" and is usually specified to be greater than 0 and less than 1,
+    /// with YCSB using 0.99, though values greater than 1 also seem to generate reasonable
+    /// distributions.
```

Comment:
Move the explanation of why `assert(theta != 1.0)` to the function's documentation.

```suggestion
    /// `theta` is the "skew" and is usually specified to be greater than 0 and less than 1,
    /// with YCSB using 0.99, though values greater than 1 also seem to generate reasonable
    /// distributions. `theta = 1` isn't allowed since it does not behave reasonably. 
```